### PR TITLE
implemented up/down mixing combinations (cf. #43)

### PR DESCRIPTION
**Pull request by b-ma**
I hope this is clean enough regarding performance and Rust conventions. Let me know if it's ok for you.

_Aside_: I wondered if renaming the argument `channels` to `computed_number_of_channels` or something, wouldn't be more clear regarding the spec.

### CODE REVIEW

**Comment by orottier** [comment]
Hi @b-ma , this looks very good!
Feel free to adjust variable names to correspond to the spec more.
Also, could you run `cargo clippy` and `cargo format` and implement the suggested changes for `src/alloc.rs` (depending on your rustc version, warnings may appear for other files too)
After that we are ready to merge!

**Comment by Jerboas86** [comment]
These are the clippy flags that i use at the top of my modules to make clippy especially picky. Fell free to use it.

```rust
// for example at the top of alloc.rs
#![warn(
    clippy::all,
    clippy::pedantic,
    clippy::nursery,
    clippy::perf,
    clippy::missing_docs_in_private_items
)]
```

be careful to false positive though

And thank you for your work.

**Comment by b-ma** [comment]
Hey, thanks for tips. I actually only used the `all` and `perf` clippy flags because the other ones gave me hints I'm really not sure I understand enough.

File: src/alloc.rs [hunk 1]
```diff
@@ -1,12 +1,20 @@
...
 
 use arrayvec::ArrayVec;
 use std::cell::RefCell;
 use std::rc::Rc;
 
 use crate::buffer::ChannelInterpretation;
 
-const LEN: usize = crate::BUFFER_SIZE as usize;
+/// size of a buffer, 128 samples per spec
+const BUFFER_SIZE: usize = crate::BUFFER_SIZE as usize;
```
Comment (b-ma) [inline src/alloc.rs:17]: maybe this is weird

Comment (orottier) [inline src/alloc.rs:17]: Not sure how I feel about 'shadowing' this const, but maybe all in all BUFFER_SIZE should just be `usize` always. I'll make a separate issue

**Pull request merged**